File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    headers = {}
    
    for phone_number in phone_book:
        headers[phone_number] = 1 #딕셔너리에 폰번호 추가
      
    
    for phone_number in phone_book: #하나의 번호에서
        header = ''
        for number in phone_number: #번호를 단위로 쪼개서
            header += number  #문자열에 추가해주면서
            if header in headers and header != phone_number: #단위로 쪼갠 번호가 딕셔너리에 있으면 return False
                return False
    return True

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    var answer = true;
    // 시간초과
    // phone_book.sort()
    // console.log(phone_book)
    // while(phone_book.length>1){
    //     let val = phone_book.shift()
    //     for (let i = 0; i<phone_book.length; i++) {
    //         //console.log(phone_book[i].slice(0,val.length+1), val)
    //         if (phone_book[i].slice(0,val.length) == val) {
    //         return false
    //             }
    //     }
    // }
    var mset = {} 
    phone_book.sort()
    phone_book.forEach(v=>{
        //console.log(v)
        for(let i =1; i<v.length; i++) {
            mset[v.slice(0,i)] = 'check'
            //console.log(mset)
        }
    })
    phone_book.forEach(v=>{
        if (v in mset) {
            answer = false
        }
    })
    return answer;
}